### DISCUSSION TITLE
Introducing Serwist.

### DISCUSSION THREAD

**Discussion by DuCanhGH** [DuCanhGH/next-pwa#121] [Announcements]
## Serwist

I'm proud to announce Serwist, the successor of `@ducanh2912/next-pwa`. Here is its current home: https://github.com/serwist/serwist. What have changed between `@ducanh2912/next-pwa@9.7.2` and `@serwist/next@8.0.0` are listed [here](https://github.com/serwist/serwist/pull/1).

It is a fork of Workbox, which is why Workbox's commit history and list of contributors remain intact. We don't, however, have it marked as a fork because it causes issues with searching for the repo on Github. Two months of work (I actually slacked off a lot  to be honest) went into it so as to remove unnecessary burden on the codebase.

It is also a successor of `@ducanh2912/next-sw`, which used to be my silly attempt at reinventing the wheel...

## Future plans

- Reintroduce a documentation website.
- Add `@serwist/vite`.
- Remove unused code.
- Re-add examples.
- Re-add tests.
- Re-add temporarily removed features.

## Why?

Maintenance for Workbox pretty much... stopped. There hasn't been any activity from its active maintainer for nearly half a year, and that last activity also took place after nearly a year of hiatus. Some of its dependencies are deprecated/vulnerable, which leaves us to a hellhole of security risks. I felt like taking the project under my wing, but, same as `@ducanh2912/next-pwa`, this project won't be advertised around.

## What does this mean for `@ducanh2912/next-pwa`?

This changes things for the package. While I won't add a lot of new features like in the past, I will keep it maintained. This repo was what truly introduced me to the world of doing open source, and it has taught me a lot of things.

Note that `@ducanh2912/next-pwa@10`, the last (or not?) major release, will be released soon (approx. 3 weeks). Being essentially a backport from `@serwist/next@8.0.0`, it will change and remove a lot of unnecessary features.

## Why was GenerateSW removed?

I felt like it was too much burden on Serwist's codebase for pretty much nothing. In exchange, I added `@serwist/sw`, which exports `installSerwist`, a function that works like GenerateSW in InjectManifest. Simply put, to migrate from `@ducanh2912/next-pwa` to `@serwist/next`, you only need to add `app/sw.{ts,js}`:

```ts
import { defaultCache } from "@serwist/next/browser";
import type { PrecacheEntry } from "@serwist/precaching";
import { installSerwist } from "@serwist/sw";

declare const self: ServiceWorkerGlobalScope & {
  // Change this attribute's name to your `injectionPoint`.
  __SW_MANIFEST: (PrecacheEntry | string)[] | undefined;
};

installSerwist({
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: defaultCache,
});
```

Then change your `next.config.mjs`:

```js
// @ts-check
import withSerwistInit from "@serwist/next";

const withSerwist = withSerwistInit({
  cacheOnFrontEndNav: true,
  swSrc: "app/sw.ts",
  swDest: "public/sw.js",
});

/** @type {import("next").NextConfig} */
const nextConfig = {
  reactStrictMode: true,
};

export default withSerwist(nextConfig);
```

This allows you to use modules in `urlPattern` since we no longer try to stringify the function.

## NodeNext support

~~While `moduleResolution: "nodenext"` with `"type": "module"` works fine, `"type": "commonjs"` doesn't for now. This is because I've not figured out a way to emit `.d.cts` for  `.ts` files with `tsc` (it would emit `.d.cts` for `.cts` files, but I'd rather not duplicate code for such a silly reason) without `rollup-plugin-dts` (this package doesn't play nice with `declarationMap` for obvious reasons). So, unfortunately, I guess you'd have to use `moduleResolution: "node"` for now (or just start using `"type": "module"`) if you are using Serwist.~~

The above issue 🤓☝️ has been fixed.

## Afterword

Thank you for your continued support! It really means a lot to me :)

I'd also like to thank @KPixel (sorry for tagging you the millionth time) for helping me out with the project! 

Happy thoughts...
![image](https://github.com/DuCanhGH/next-pwa/assets/75556609/16adbcf3-6129-44ac-9a0d-af4c24b7b9cc)

**Comment by SahilAggarwal2004** [comment]
Although serwist is good, but there are some features that are available in your next-pwa fork but not in @serwist/next.
Can we expect those features in serwist or should I stick with next-pwa?

Edit: The feature I want the most is offline fallbacks. I just read that you have removed it temporarily.

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 some features are permanently removed, but some are only gone temporarily. Those that I removed permanently simply don't matter anymore, to be honest. As for the rest, they are planned to be re-added, but it's possible that some, too, will be removed with no return in sight. `fallbacks` won't though, because the reason I removed it is just that I'm investigating a better way of doing it, since `@serwist/next` doesn't have access to `runtimeCaching`.

If there's demand for a permanently removed feature, I guess you'd have no choice but to find an alternative for it or use `next-pwa`, which remains maintained.

**Reply by SahilAggarwal2004** [reply]
Okay, thanks for confirming!

**Reply by SahilAggarwal2004** [reply]
How can I configure `offlineAnalyticsConfig`? As in next-pwa, I just had to set `offlineGoogleAnalytics` to true.

**Reply by DuCanhGH** [reply]
> How can I configure `offlineAnalyticsConfig`? As in next-pwa, I just had to set `offlineGoogleAnalytics` to true.

@SahilAggarwal2004 ah crap, seems that we don't allow booleans right now. Will recheck soon!

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 fixed in 8.0.2 :)

**Reply by SahilAggarwal2004** [reply]
Woah, that was quick! Also good to see that you removed multiple useless dependencies.

Thanks for the update!

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 `depcheck` is one hell of a package. Very neat :)
Also, you probably mean "unused", not "useless" 💀

**Reply by SahilAggarwal2004** [reply]
Hmm, I will try `depcheck` in my projects too.

Learned a lot from you today. Thanks for sharing your knowledge!

And yeah, I meant "unused" 😂😅

**Reply by SahilAggarwal2004** [reply]
Did you figure out the offline fallbacks thing? Although its not that big issue, but still waiting for it. I understand that you are currently busy building the documentation and other important things, and I appreciate that.
Please just let me know on this discussion when you figure it out and finally implement it.

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 there was a prototype, but it involved too much garbage and messed with Webpack's caching... That's why I'm still thinking about how this feature should be designed.

**Reply by SahilAggarwal2004** [reply]
How are things setup in the next-pwa library? Can't the same model be applied here?

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 the same mechanism can be applied for Serwist, but I'm thinking about something better. This is because the old mechanism requires modifying `runtimeCaching` according to the `fallbacks` object, and that is not something we can do in `@serwist/next`, which means that `@serwist/sw` has to do that job, but `@serwist/next` also has to know the `fallbacks` object to modify `manifestEntries` (we want to use a revision for these `fallbacks` URLs, and I want to create that revision during the build). All that means we have to inject yet another value into `swSrc`, this value being the `fallbacks` object. That's not what I want the feature to look like. I want something more in line with the introduction of `@serwist/sw`. If you want to use offline fallbacks right now though, there is an alternative:
```js
import withSerwistInit from "@serwist/next";

const withSerwist = withSerwistInit({
  cacheOnFrontEndNav: true,
  swSrc: "app/sw.ts",
  swDest: "public/sw.js",
  additionalPrecacheEntries: ["/fallback-font.ttf"],
});
```

```ts
installSerwist({
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: [
    {
      urlPattern: /^https:\/\/fonts\.(?:gstatic)\.com\/.*/i,
      handler: "CacheFirst",
      options: {
        cacheName: "google-fonts-webfonts",
        expiration: {
          maxEntries: 4,
          maxAgeSeconds: 365 * 24 * 60 * 60, // 365 days
        },
        plugins: [
          {
            async handlerDidError() {
              const fallbackResponse = await caches.match("/fallback-font.ttf");
              if (fallbackResponse !== undefined) {
                return fallbackResponse;
              }
              return Response.error();
            },
          },
        ],
      },
    },
    ...defaultCache.filter((entry) => !(entry.options?.cacheName === "google-fonts-webfonts")),
  ],
});
```

**Reply by SahilAggarwal2004** [reply]
Thanks a lot for providing the alternative! I will implement it for the time being 🤗

**Reply by SahilAggarwal2004** [reply]
Well, how about using a separate config file dedicated for serwist?
Something like tailwind does

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 that config file is this wonderful `installSerwist` function :) I actually planned to introduce a config file (using the same one used by `@serwist/cli`), but I hit upon this bright idea of doing this instead. Well, the biggest drawback is the one we are facing right now to be honest. We can't really introduce yet another config file though, for we already have 2 (one being `swSrc`, and the other being `next.config.js`).

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 the API that I imagine is something like this:

- The `fallbacks` option now belongs to `installSerwist` and is an array of objects, each object of which now uses a `matcher`, allowing you to use all criterias to match the resource rather than restricting you to `request.destination`, and requires you to provide a revision in addition to an URL, allowing `@serwist/precaching` to precache these URLs correctly.

How does that sound to you as an user? I think this is the best way of doing offline fallbacks :) It is much more framework agnostic than the old mechanism of `next-pwa` to be honest, for the functionality now belongs to `@serwist/sw` rather than `@serwist/next`.

**Reply by SahilAggarwal2004** [reply]
Hmm, valid point. But I was proposing if `next.config.js` could be eliminated for the new config file. Please ignore if this idea is stupid as I don't have a good knowledge about how config files actually work.

And btw can you suggest a regex for just page routes? or any way to dynamically add hostname in regex or hardcoding it is the only way?

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 that's a valid point, but that right there is a breaking change. I'd avoid doing something like that so soon to be honest, since there's no reason to do that right now. Also, speaking about the implementation, it'd be difficult to support TypeScript configs, as we would have to transpile it before importing it. So, that's a no-go (at least for now).

I don't understand the question though.

**Reply by SahilAggarwal2004** [reply]
The alternative solution you provided was not working for any regex. Then I tried to allow it for everything, so I just did:

```
{
    urlPattern: () => true,
    handler: 'CacheFirst',
    options: {
        cacheName: 'fallback',
        plugins: [{
            handlerDidError: async () => {
                const fallbackResponse = await caches.match('/_offline');
                if (fallbackResponse !== undefined) return fallbackResponse
                return Response.error();
            }
        }]
    }
}
```
But it's still not working :(

**Reply by SahilAggarwal2004** [reply]
> @SahilAggarwal2004 that's a valid point, but that right there is a breaking change. I'd avoid doing something like that so soon to be honest, since there's no reason to do that right now. Also, speaking about the implementation, it'd be difficult to support TypeScript configs, as we would have to transpile it before importing it. So, that's a no-go (at least for now).
> 
> I don't understand the question though.

Yeah I understand your concern there. It's better to keep the shift process from next-pwa to serwist as minimal as possible

**Reply by SahilAggarwal2004** [reply]
> @SahilAggarwal2004 the API that I imagine is something like this:
> 
> - The `fallbacks` option now belongs to `installSerwist` and is an array of objects, each object of which now uses a `matcher`, allowing you to use all criterias to match the resource rather than restricting you to `request.destination`, and requires you to provide a revision in addition to an URL, allowing `@serwist/precaching` to precache these URLs correctly.
> 
> How does that sound to you as an user? I think this is the best way of doing offline fallbacks :) It is much more framework agnostic than the old mechanism of `next-pwa` to be honest, for the functionality now belongs to `@serwist/sw` rather than `@serwist/next`.

Currently, my requirement is very basic. So I am okay with both the mechanisms. But in case of complex requirements (which many developers may have), your approach seems to be better.

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 is "/_offline" precached? Wondering what could cause it to not work. Though, I must clarify that while it should work hypothetically, I've not tried it out myself. Perhaps you should try adding `{ ignoreSearch: true }` to the `caches.match()` part. Serwist does insert a search param into some of the URLs it precaches, after all.

**Reply by SahilAggarwal2004** [reply]
Yup "/_offline" is precached, I even checked the Caches list in application tab to double check.
I will try your suggested change tomorrow as I am going to bed now

**Reply by SahilAggarwal2004** [reply]
The code inside the block is just never running, I did a `console.log(fallbackResponse, await caches.keys())` to check what's happening, but nothing is being logged on console.

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 that's weird. The reproduction code works for me:
![image](https://github.com/DuCanhGH/next-pwa/assets/75556609/5e8c1535-0484-41bc-8d24-3cb281331a67)

**Reply by SahilAggarwal2004** [reply]
Wait, let me try by removing all other runtime caching patterns

**Reply by SahilAggarwal2004** [reply]
Ohk, found another pattern in runtime caching which was matching with the route on which I was testing offline fallback. It was the reason why handlerDidError code wasn't executing...😅

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 alright 🤨 Though I might have to add that if you place your list of custom entries above the default `runtimeCaching` array, the former will take precedence over the latter.

**Reply by SahilAggarwal2004** [reply]
Yeah, I knew about it that's why I thought to remove all other patterns as this pattern was at the end of the array. Thanks though, it worked now!
Can you suggest a urlPattern, either regex or a function for just page routes? As I don't want to serve any api call or any other unexpected request to be served with `/_offline`.

**Reply by SahilAggarwal2004** [reply]
For now, I am comparing `request.destination === 'document'`

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 using `request.destination === 'document'` is fine :) Wondering if that works with RSC though.

**Reply by SahilAggarwal2004** [reply]
I am not using RSC as of now, just because of PWAs.
The biggest problem there is not cache creation but cache invalidation.
Suppose a user logs into a website, and I cache his/her user specific data for better offline experience. Then he/she decides to log out. Currently, I have no mechanism to clear that cache and if some other user wants to see the data, it will easily be accessible. Whereas in a client component, what I do is pre-cache a general page and use localstorage to save and access user specific data and if someone decides to logout, I just have to clear the localstorage.

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 eh, you do you. I have never had any problem of the sort to be honest. Perhaps you can consider using NetworkOnly for authentication-related stuff and adding a fallback for when the user is offline?

**Reply by SahilAggarwal2004** [reply]
There is a problem in that approach is that the offline experience is not that great.
Suppose I built a note-taking application. Now, if the user is not online, he won't be able to even see his notes, instead he will just see a page that he is offline. Whereas in the approach I am using, the notes are available irrespective of the network connection, with the only drawback that the user might see stale notes in case he is offline and notes were updated from some other device. But common, its far better than showing the "You are Offline" page!

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 eh, if the notes are available offline, they are just straight up visible to any other user (you can access the pages through Cache Storage). I guess that's just the shitty tradeoff you've got to make for offline support. Perhaps it's better to use something like the Authorization header (consisting of a token, also add this header to the Vary header through the middleware) and check for it in the page's loader before accessing the notes' data. This is just a hypothesis though, and the implementation might look a tad bit too different...

**Reply by SahilAggarwal2004** [reply]
I think you didn't get my point. The notes are not stored in cache storage but in localstorage, and as soon as someone logs out, I simply clear them from localstorage. The page stored in cache storage is just a general page containing the code to access the notes from the localstorage.

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 I assumed a normal authenication system with a database. If fetching from localStorage is your way of doing it though, that's fine :)

**Reply by SahilAggarwal2004** [reply]
Yeah, actually there's a database call too, and when the notes are fetched, I update them in localstorage

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 sorry for making you wait. Here is `fallbacks`, implemented in 8.1.0:
```js
installSerwist({
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: defaultCache,
  fallbacks: {
    entries: [
      {
        url: "/~offline",
        revision,
        matcher({ request }) {
          return request.destination === "document";
        },
      },
    ],
  },
});
```
Pretty neat, I'd say :)

**Reply by SahilAggarwal2004** [reply]
Thanks a lot! Tbh, I was expecting that it will take more time considering the fact that you are busy making the documentation too.

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 eh, the docs doesn't hold a priority higher than the features themselves :) It's just that I was struggling with coming up with a better API.

**Reply by SahilAggarwal2004** [reply]
So how did you figure things out after all?

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 eh, it was a combination of old ideas and an eureka moment :)

**Reply by SahilAggarwal2004** [reply]
Btw is there a clean way to checkout the changes in different versions? It's hard to track changes in a monorepo

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 I'm afraid that we don't have anything of the sort right now. However, perhaps we can start doing what Next.js has been doing - adding changelogs to its documentation site. Thinking about that :)

**Reply by SahilAggarwal2004** [reply]
Yeah that would be great!

**Reply by maiconcarraro** [reply]
I'm usually trying to find the changes by doing this search: https://github.com/serwist/serwist/releases?q=%40serwist%2Fsw%40&expanded=true

**Reply by SahilAggarwal2004** [reply]
@maiconcarraro Thanks for the recommendation!

**Reply by SahilAggarwal2004** [reply]
Hey @DuCanhGH,
I was trying out server actions, and there is a bug I am not able to resolve. The issue is that if the server action call fails because of any reason (mostly when the user is offline), nextjs throws an error on the screen. I tried encapsulating the call in a try, catch block but it doesn't help. Is there any way to handle fallbacks for server action calls built into serwist or I have to use an error boundary?

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 mate this is not the place to ask these bloody questions. Just use Serwist's Discussions.

Also, `installSerwist`'s `fallbacks` should just work. It's just that you have to figure out a way to match requests to those server actions. Perhaps there are some request headers that determine whether a Request is one to a server action.

**Reply by SahilAggarwal2004** [reply]
Okay thanks! And sorry, I will keep this in mind to ask queries on serwist discussions instead.

**Reply by SahilAggarwal2004** [reply]
It's a post request, how do I enable post request in runtime caching?

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 use `RuntimeCaching.method`.

**Reply by SahilAggarwal2004** [reply]
Thanks again!

**Reply by SahilAggarwal2004** [reply]
Unfortunately, it's not working. I don't know why, I am using the below runtime caching but still "GET" requests are being logged.
```
runtimeCaching: [
        {
            urlPattern: (req) => {
                console.log(req)
                return true
            },
            handler: 'NetworkFirst',
            method: 'POST',
            options: { cacheName: 'server-action' }
        },
        {
            urlPattern: matcher,
            handler: 'NetworkOnly',
            options: { cacheName: 'documents' }
        },
        {
            urlPattern: /\.(?:eot|otf|ttc|ttf|woff|woff2|font.css)$/i,
            handler: 'CacheFirst',
            options: { cacheName: 'static-font-assets' }
        },
        {
            urlPattern: /\.(?:jpg|jpeg|gif|png|svg|ico|webp)$/i,
            handler: 'CacheFirst',
            options: { cacheName: 'static-image-assets' }
        },
        {
            urlPattern: /\/_next\/image\?url=.+$/i,
            handler: 'CacheFirst',
            options: { cacheName: 'next-image' }
        },
        {
            urlPattern: /\.(?:mp3|wav|ogg)$/i,
            handler: 'CacheFirst',
            options: {
                rangeRequests: true,
                cacheName: 'static-audio-assets'
            }
        },
        {
            urlPattern: /\.(?:mp4)$/i,
            handler: 'CacheFirst',
            options: {
                rangeRequests: true,
                cacheName: 'static-video-assets'
            }
        },
        {
            urlPattern: /\.(?:js)$/i,
            handler: 'StaleWhileRevalidate',
            options: { cacheName: 'js-assets' }
        },
        {
            urlPattern: /\.(?:css|less)$/i,
            handler: 'CacheFirst',
            options: { cacheName: 'static-style-assets' }
        },
        {
            urlPattern: /\/_next\/data\/.+\/.+\.json$/i,
            handler: 'StaleWhileRevalidate',
            options: { cacheName: 'next-data' }
        },
        {
            urlPattern: /\.(?:json|xml|csv)$/i,
            handler: 'NetworkFirst',
            options: { cacheName: 'static-data-assets' }
        },
        {
            urlPattern: /^\/geocheck/,
            handler: 'NetworkFirst',
            options: { cacheName: 'location' }
        },
        {
            urlPattern: () => true,
            handler: 'NetworkOnly',
            options: { cacheName: 'others' }
        }
    ]
```
console.log() is only in the function in which method is set to POST

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 bloody hell. Just found out that we forgot to handle `method` for `handler` of type string...

**Reply by SahilAggarwal2004** [reply]
I don't know how to react to this situation. Although I am stuck but this still seems funny, more because it is very much relatable to me (I do these type of silly mistakes very often).
Anyways, I am happy that I was able to contribute to the library indirectly.

**Reply by SahilAggarwal2004** [reply]
I see that the issue has now been resolved in `v8.4.3`. Thanks a lot for that!
But when I log the request on console, there is no header info, it's an empty object for all the requests. Do I need to configure something so that I can receive the headers too?

**Reply by SahilAggarwal2004** [reply]
Also, I don't know why but `Link` component imported from `next/link` doesn't work for the first time when client goes offline and navigates perfectly on further tries. I am not able to figure out what could be the reason for this

**Reply by DuCanhGH** [reply]
@SahilAggarwal2004 try adding `.entries()` to the log.

> Also, I don't know why but Link component imported from next/link doesn't work for the first time when client goes offline and navigates perfectly on further tries. I am not able to figure out what could be the reason for this

You have to visit a page once before it is available offline. If that's not what you want, precache those pages.

**Comment by maiconsanson** [comment]
@DuCanhGH the URL of https://github.com/serwist/serwist is going to https://www.youtube.com/watch?v=dQw4w9WgXcQ 😄 

Maybe this little migration guide could be in a place more visible.
Thank you for your effort, by the way.

**Reply by DuCanhGH** [reply]
@maiconsanson oops, called out 🫢 Removing that :D

**Comment by MatsG23** [comment]
I cannot install this package. It is like only macOS had been specifically targeted yet.
```
npm ERR! code EBADPLATFORM
npm ERR! notsup Unsupported platform for fsevents@2.3.3: wanted {"os":"darwin"} (current: {"os":"linux"})
npm ERR! notsup Valid os:  darwin
npm ERR! notsup Actual os: linux
```

**Reply by DuCanhGH** [reply]
@MatsG23 I daily drive Linux 🤨 Seems that everything works fine on my end, which is weird. Can you try other package managers such as pnpm, bun, and yarn? Perhaps they don't enforce this OS requirement.

Edit: try 8.0.3, in which I removed fsevents! Seems that it isn't even used anyway.

**Reply by maiconsanson** [reply]
I've migrated without problems from `"@ducanh2912/next-pwa": "9.7.2"` to `"@serwist/next": "8.0.2"` using `Ventura 13.6` and `pnpm 8.11.0` at **M1 PRO**.

**Reply by DuCanhGH** [reply]
@maiconsanson well, this problem specifically does not happen to macOS users 😅 It's the opposite, actually. Not that I could reproduce that, but `fsevents` does indeed have an `"os"` array in its package.json.

**Reply by maiconsanson** [reply]
Sure! Now I see that I was dumb. 😄 Sorry for the hurry.

**Reply by MatsG23** [reply]
Works with version 8.0.3 without fsevents! Thanks.

**Comment by Murkrage** [comment]
I'm not sure if this is good or not, haven't looked into it enough yet, but I really appreciate your continued effort! Would you recommend to continue using `@serwist/sw` or is there an alternative to get PWA's working for Next?

**Reply by DuCanhGH** [reply]
@Murkrage obviously I would recommend using Serwist, but if that doesn't work, there are other alternatives such as `@ducanh2912/next-pwa` (it remains supported by the way), Workbox, `vite-plugin-pwa` (if you want to migrate from Next.js to Vite, that is), and `next-pwa` (it is pretty much abandoned though).

**Reply by Murkrage** [reply]
Oh yeah no I mostly mean the way we configure Next. Is the recommended way still to use the Next configuration file together with the generate / install method?

**Reply by DuCanhGH** [reply]
@Murkrage I actually lost you there 😅 Seems that you are not very familiar with English. But if I inferred this right, then no, Serwist doesn't have GenerateSW, which means that there's no plug-n-play way of using it. `@serwist/sw` exists to fill its vacant spot, but using this package still obviously involves a bit of code. You can use InjectManifest like how you would with Workbox though, so that's an alternative.

**Reply by Murkrage** [reply]
> Seems that you are not very familiar with English

What an odd thing to say... I'm a native speaker but I'm not sure how that's relevant 🙃

I'm indeed referring to a plug-and-play solution like `next-pwa`, or now `serwist/sw`. Whether the plug-and-play solution is the recommended way, which comes with a support burden on your end, or if we should try to move away from that solution and instead work with the core set of tools Serwist, like Workbox, provides.

**Reply by DuCanhGH** [reply]
@Murkrage huh, sorry for that then. I assumed that because I noticed a grammar error and you seemed to struggle with describing your question (I actually didn't understand "generate / install method")... Perhaps it's actually the native speakers that don't emphasize on the importance of grammar, which has been grilled into my head after having learnt English for too long. Putting this problem aside, `@serwist/sw` is the abstraction on top of Serwist's tools, replacing GenerateSW, so you can think of it as the "plug-and-play" way of using Serwist. How you would use it is described in the post :)

**Comment by ajayvignesh01** [comment]
Somehow, I wasn't able to get your fork of next-pwa working, but this is working like a charm lol.

**Reply by DuCanhGH** [reply]
💀
I'd like you to elaborate further if possible :D

**Reply by SahilAggarwal2004** [reply]
Woah, usually the reverse happens😂

**Comment by maiconcarraro** [comment]
Just migrated from next-pwa to serwist few hours ago, looking great so far, do you think about enabling the sponsor for the new project?

**Reply by DuCanhGH** [reply]
@maiconcarraro  Well, this problem has been brought up quite a few times, so I'm just gonna be frankly honest about the biggest problem, being that I don't even have a credit card, which is required for foreign services like GitHub Sponsors. I actually have a domestic banking account (I use it to buy most things, including games on Steam), but that isn't supported obviously. I'm thinking about acquiring one though. Workbox's codebase actually had me wanting to kick myself in the nuts a few times (not as badly as back when I was actively contributing to Next.js though), so that whole open source spirit was straight up thrown out of the window 💀

**Reply by maiconcarraro** [reply]
gotcha, in case you have an amazon wishlist or steam wishlist, I'm happy to buy you something as well

**Reply by maniraj1234** [reply]
Ig you can use paypal @DuCanhGH

**Comment by aNyMoRe0505** [comment]
just come here to say thank you for your effort :heart:

**Comment by aNyMoRe0505** [comment]
Hello, I'd like to confirm if my understanding is correct on the following points:

1. `@serwist/next` only supports versions of Next.js that are v14 and above, and it specifically works with app routes (not compatible with page routes).

2. If we want to use a custom worker, we just need to write the logic under `app/sw.js`.

Thank you.

**Reply by DuCanhGH** [reply]
@aNyMoRe0505 what makes you think that the package only supports App Router? The Getting started page specifically includes `pages/_app.tsx` templates. `swSrc: "app/sw.ts"` was only used to make an example. You can set this value to anything that points to a valid file, such as `"service-worker/index.ts"`.

**Reply by aNyMoRe0505** [reply]
@DuCanhGH 
sorry, I read too quickly and I haven't started the migration process yet.
I thought sw.js could only be placed in the app folder
I apologize for the misunderstanding. I'll take the time to properly migrate to serwist.
Thank you.

**Comment by maniraj1234** [comment]
Hey @DuCanhGH,
This migration guide is pretty seamless. however I'm getting a linting issue with sw.ts which I couldn't resolve :( 
5:21  error  'ServiceWorkerGlobalScope' is not defined  no-undef

**Reply by DuCanhGH** [reply]
```jsonc
{
  // Other stuff...
  "compilerOptions": {
    // Other options...
    "types": [
      // Other types...
      // This allows Serwist to type `window.serwist`.
      "@serwist/next/typings"
    ],
    "lib": [
      // Other libs...
      // Add this! Doing so adds WebWorker and ServiceWorker types to the global.
      "webworker"
    ],
  },
}
```

**Reply by maniraj1234** [reply]
still same issue. PR ref: https://github.com/maniraj1234/kuberviz/pull/13/files

**Reply by DuCanhGH** [reply]
@maniraj1234 just ESLint things. Simply disable no-undef.

See https://eslint.org/docs/latest/rules/no-undef#handled_by_typescript. This rule is not recommended for TS projects.

**Reply by maniraj1234** [reply]
This worked. thanks.

**Comment by DuCanhGH** [comment]
Please stop using this thread to ask questions related to Serwist. You should open a new Discussion in Serwist's repo instead.
